test/suites/azuregarbagecollection/suite_test.go

```diff
@@ -0,0 +1,120 @@
+)
+
+var env *azure.Environment
+var nodeClass *v1alpha2.AKSNodeClass
+var nodePool *karpv1.NodePool
+
+func TestGC(t *testing.T) {
+	RegisterFailHandler(Fail)
+	BeforeSuite(func() {
+		env = azure.NewEnvironment(t)
+	})
+	RunSpecs(t, "Azure Garbage Collection")
+}
+
+var _ = BeforeEach(func() {
+	env.BeforeEach()
+	nodeClass = env.DefaultAKSNodeClass()
+	nodePool = env.DefaultNodePool(nodeClass)
+})
+var _ = AfterEach(func() { env.Cleanup() })
+var _ = AfterEach(func() { env.AfterEach() })
+
+var _ = Describe("gc", func() {
+	It("should garbage collect network interfaces created by karpenter", func() {
+		vnet, err := getVNET(env.Context, env.VNETClient, env.VNETResourceGroup)
+		Expect(err).ToNot(HaveOccurred())
+		nicName := "orphan-nic"
+
+		err = createOrphanNIC(env.Context, env.InterfacesClient, env.NodeResourceGroup, env.Region, nicName, vnet.Properties.Subnets[0])
+		Expect(err).ToNot(HaveOccurred())
+		EventuallyExpectOrphanNicsToBeDeleted(env, env.NodeResourceGroup, nicName, env.InterfacesClient)
+	})
+})
+
+func getVNET(ctx context.Context, client *armnetwork.VirtualNetworksClient, vnetRG string) (*armnetwork.VirtualNetwork, error) {
+	pager := client.NewListPager(vnetRG, nil)
+	for pager.More() {
+		resp, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to list virtual networks: %w", err)
+		}
+		if len(resp.VirtualNetworkListResult.Value) > 0 {
+			return resp.VirtualNetworkListResult.Value[0], nil
+		}
+	}
+	return nil, fmt.Errorf("no virtual networks found in resource group: %s", vnetRG)
+}
+
+func createOrphanNIC(ctx context.Context, client *armnetwork.InterfacesClient, resourceGroup, region, nicName string, subnet *armnetwork.Subnet) error {
+	nodepoolKey := strings.ReplaceAll(karpv1.NodePoolLabelKey, "/", "_")
+	poller, err := client.BeginCreateOrUpdate(ctx, resourceGroup, nicName, armnetwork.Interface{
+		Location: to.Ptr(region),
+		Tags:     map[string]*string{nodepoolKey: lo.ToPtr("default")},
+		Properties: &armnetwork.InterfacePropertiesFormat{
+			IPConfigurations: []*armnetwork.InterfaceIPConfiguration{
+				{
+					Name: lo.ToPtr(nicName),
+					Properties: &armnetwork.InterfaceIPConfigurationPropertiesFormat{
+						Primary:                   lo.ToPtr(true),
+						Subnet:                    subnet,
+						PrivateIPAllocationMethod: lo.ToPtr(armnetwork.IPAllocationMethodDynamic),
+					},
+				},
+			},
+		},
+	}, nil)
+	if err != nil {
+		return err
+	}
+	_, err = poller.PollUntilDone(ctx, nil)
+	return err
+}
+
+func EventuallyExpectOrphanNicsToBeDeleted(env *azure.Environment, resourceGroup, nicName string, nicClient *armnetwork.InterfacesClient) {
+	GinkgoHelper()
+	Eventually(func() bool {
+		_, err := nicClient.Get(env.Context, resourceGroup, nicName, nil)
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 404 {
+			return true
+		}
+		return false
+	}, 15*time.Minute, 15*time.Second).Should(BeTrue())
```

Comment:
15 minutes feels like a long time.

The nic garbage collector is on a 5 minute loop. Maybe we could do something like 8-10 minute wait here, to have some, but less buffer?

Reply:
I hit some nrp throttling sometimes when it was at 7 minutes so left it at 15 minutes to give it 3 retries and make it less flakey.